src/librustc/session/mod.rs

```diff
@@ -336,6 +336,10 @@ impl NodeIdAssigner for Session {
     fn peek_node_id(&self) -> NodeId {
         self.next_node_id.get().checked_add(1).unwrap()
     }
+
+    fn span_fatal(&self, sp: MultiSpan, msg: &str) -> ! {
```

Comment:
NodeIdAssigner could have a GetDiagnostic method returning Option<Diagnostic>, that avoids having error methods for every kind of error in NodeIdAssigner.